File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/AbstractOverlay.java
package io.github.chindeaone.collectiontracker.gui.overlays;

import io.github.chindeaone.collectiontracker.config.core.Position;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Font;
import net.minecraft.client.gui.GuiGraphics;
import java.util.Collections;
import java.util.List;

/**
 * A general abstract class for all overlays.
 */
public abstract class AbstractOverlay {
    protected boolean renderingAllowed = true;

    public abstract String overlayLabel();

    public abstract Position position();

    public abstract boolean isEnabled();

    public boolean isRenderingAllowed() {
        return renderingAllowed;
    }

    public void setRenderingAllowed(boolean allowed) {
        this.renderingAllowed = allowed;
    }

    public boolean shouldRender() {
        return isEnabled() && isRenderingAllowed();
    }

    public abstract void render(GuiGraphics context);

    public abstract void updateDimensions();

    public void handleLineAction(String line) {}

    public List<String> getLines() {
        return Collections.emptyList();
    }

    public boolean handleMouseClick(double mouseX, double mouseY) {
        if (!isEnabled() || !isHovered(mouseX, mouseY)) return false;

        List<String> lines = getLines();
        if (lines.isEmpty()) return false;

        Position position = this.position();
        int x = position.getX();
        int y = position.getY();
        float scale = position.getScale();

        Font fr = Minecraft.getInstance().font;
        int height = (int) (fr.lineHeight * lines.size() * scale);
        int width = (int) (position.getWidth() * scale);

        if (mouseX >= x && mouseX <= x + width && mouseY >= y && mouseY <= y + height) {
            double relativeY = mouseY - y;
            int lineClicked = (int) (relativeY / (fr.lineHeight * scale));

            if (lineClicked >= 0 && lineClicked < lines.size()) {
                handleLineAction(lines.get(lineClicked));
                return true;
            }
        }
        return false;
    }

    public boolean isHovered(double mouseX, double mouseY) {
        Position pos = this.position();
        if (pos == null) return false;

        int yPadding = 4;
        int x = pos.getX();
        int y = pos.getY();
        float s = pos.getScale();
        int w = Math.round(pos.getWidth() * s);

        double x2 = x + w;
        double y1 = y - (yPadding * s);
        double y2 = y + (pos.getHeight() + yPadding) * s;

        return mouseX >= x && mouseX <= x2 && mouseY >= y1 && mouseY <= y2;
    }
}

File: src/main/java/io/github/chindeaone/collectiontracker/gui/overlays/AbstractOverlay.kt
package io.github.chindeaone.collectiontracker.gui.overlays

import io.github.chindeaone.collectiontracker.config.core.Position
import io.github.chindeaone.collectiontracker.gui.GuiManager.getEditorInstance
import io.github.chindeaone.collectiontracker.gui.GuiManager.openEditor
import io.github.chindeaone.collectiontracker.utils.HypixelUtils
import io.github.chindeaone.collectiontracker.utils.rendering.RenderUtils
import io.github.notenoughupdates.moulconfig.gui.MoulConfigEditor
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import kotlin.math.roundToInt

/**
 * A general abstract class for all overlays
 */
abstract class AbstractOverlay {
    var isRenderingAllowed: Boolean = true

    abstract val overlayLabel: String

    abstract val position: Position

    open val isEnabled: Boolean get() = false

    fun shouldRender(): Boolean {
        val mc = Minecraft.getInstance()
        return !mc./*? if 26.2 {*/ /*gui.hud.isHidden() */ /*?} else {*/options.hideGui /*?}*/
                && !mc.debugEntries.isOverlayVisible
                && isEnabled
                && isRenderingAllowed
    }

    open fun render(context: GuiGraphicsExtractor) {
        if (!isEnabled) return

        val lines = lines
        if (lines.isEmpty()) return

        RenderUtils.drawOverlayFrame(context, position) { RenderUtils.renderStrings(context, lines) }
    }

    open fun updateDimensions() {
        if (!HypixelUtils.isInSkyblock && !isEnabled) return

        val lines = lines
        if (lines.isEmpty()) {
            position.setDimensions(0, 0)
            return
        }

        val fr = Minecraft.getInstance().font
        var maxW = 0
        for (l in lines) maxW = maxOf(maxW, fr.width(l))
        val h = fr.lineHeight * lines.size

        position.setDimensions(maxW, h)
    }

    open val lines: List<String> get() = emptyList()

    open fun handleLineAction(line: String) {}

    fun handleMouseClick(mouseX: Double, mouseY: Double): Boolean {
        if (!isEnabled || !isHovered(mouseX, mouseY)) return false

        val lines = lines
        if (lines.isEmpty()) return false

        val position = position
        val x = position.x
        val y = position.y
        val scale = position.scale

        val fr = Minecraft.getInstance().font
        val height = (fr.lineHeight * lines.size * scale).toInt()
        val width = (position.width * scale).toInt()

        if (mouseX in x.toDouble()..(x + width).toDouble() && mouseY in y.toDouble()..(y + height).toDouble()) {
            val relativeY = mouseY - y
            val lineClicked = (relativeY / (fr.lineHeight * scale)).toInt()

            if (lineClicked in lines.indices) {
                handleLineAction(lines[lineClicked])
                return true
            }
        }
        return false
    }

    open fun isHovered(mouseX: Double, mouseY: Double): Boolean {
        val pos = position

        val padding = 4

        val x = pos.x
        val y = pos.y
        val scale = pos.scale

        val width = (pos.width * scale).roundToInt()
        val height = (pos.height * scale).roundToInt()

        val x2 = (x + width).toDouble()
        val y1 = (y - padding * scale).toDouble()
        val y2 = (y + height + padding * scale).toDouble()

        return mouseX in x.toDouble()..x2 && mouseY in y1..y2
    }

    fun jumpToConfig() {
        val pos = position

        val editor: MoulConfigEditor<*> = getEditorInstance()
        val option = editor.getOptionFromField(pos.link) ?: return

        editor.search("")
        if (!editor.goToOption(option)) return
        openEditor(editor)
    }
}
